providers/vultr/provider.go

```diff
@@ -0,0 +1,242 @@
+	userdataString, err := engine.RenderUserDataTemplate(p.config, agent, p.userData)
+	if err != nil {
+		return fmt.Errorf("%s: RenderUserDataTemplate: %w", p.name, err)
+	}
+
+	image := -1
+	osList, _, _, err := p.client.OS.List(ctx, &govultr.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("%s: OS.List: %w", p.name, err)
+	}
+	for _, osS := range osList {
+		if osS.Name == p.image {
+			image = osS.ID
+			break
+		}
+	}
+	if image == -1 {
+		return fmt.Errorf("%s: DeployAgent: no image found for %s", p.name, p.image)
+	}
+	tags := make([]string, 0)
+	for key, item := range p.labels {
+		tags = append(tags, fmt.Sprintf("%s=%s", key, item))
+	}
+
+	instance, _, err := p.client.Instance.Create(ctx, &govultr.InstanceCreateReq{
+		Hostname:        agent.Name,
+		UserData:        base64.StdEncoding.EncodeToString([]byte(userdataString)),
+		Plan:            p.plan,
+		Region:          p.region,
+		Label:           agent.Name,
+		Tags:            tags,
+		OsID:            image,
+		EnableVPC:       govultr.BoolToBoolPtr(false), // TODO: allow to use private networks
+		ActivationEmail: govultr.BoolToBoolPtr(false),
+		SSHKeys:         p.sshKeys,
+		EnableIPv6:      &p.enableIPv6,
+	})
+	if err != nil {
+		return fmt.Errorf("%s: Server.Create: %w", p.name, err)
+	}
+
+	// TODO: move to provider utils and use backoff?
+	log.Debug().Msgf("waiting for instance %s", instance.ID)
+	for range 5 {
+		agents, err := p.ListDeployedAgentNames(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to return list for agents")
+		}
+
+		for _, a := range agents {
+			if a == agent.Name {
+				return nil
+			}
+		}
+
+		log.Debug().Msgf("created agent not found in list yet")
+		time.Sleep(1 * time.Second)
+	}
+
+	return fmt.Errorf("instance did not resolve in agent list: %s", instance.ID)
+}
+
+func (p *Provider) getAgent(ctx context.Context, agent *woodpecker.Agent) (*govultr.Instance, error) {
+	servers, _, _, err := p.client.Instance.List(ctx, &govultr.ListOptions{
+		Label: agent.Name,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", p.name, err)
+	}
+
+	if len(servers) == 0 {
+		return nil, nil
+	}
+
+	if len(servers) > 1 {
+		return nil, fmt.Errorf("%s: getAgent: found multiple instances with label %s", p.name, agent.Name)
+	}
+
+	return &servers[0], nil
+}
+
+func (p *Provider) RemoveAgent(ctx context.Context, agent *woodpecker.Agent) error {
+	server, err := p.getAgent(ctx, agent)
+	if err != nil {
+		return fmt.Errorf("%s: %w", p.name, err)
+	}
+
+	if server == nil {
+		return nil
+	}
+
+	err = p.client.Instance.Delete(ctx, server.ID)
+	if err != nil {
+		return fmt.Errorf("%s: %w", p.name, err)
+	}
+
+	return nil
+}
+
+func (p *Provider) ListDeployedAgentNames(ctx context.Context) ([]string, error) {
+	var names []string
+	pageOpts := 200
+	listOptions := &govultr.ListOptions{
+		Tag:     engine.LabelPool + "=" + p.config.PoolID,
+		PerPage: pageOpts,
+	}
+
+	servers, _, _, err := p.client.Instance.List(ctx,
```

Comment:
Could this use proper pagination in case someone has >200 instances?

Reply:
Added